[WIP] feature: doc restructure (#21)
yongenaelf/aelf-docs: docs/tutorial/stackup/2. smart-contract/2.3 Implement-Smart-Contract-Logic.md

---
sidebar_position: 3
---

# Implement Voting Smart Contract Logic

> In this step, we will be implementing the logic of our voting smart contract in the **src/BuilderspDAO.cs** file!

Reviewer: ```suggestion
In this step, we will be implementing the logic of our voting smart contract in the **src/BuildersDAO.cs** file!
```

using AElf.Types;
using Google.Protobuf.WellKnownTypes;
using System.Collections.Generic;
using System.Security.Principal;

namespace AElf.Contracts.BuildersDAO
{
    public class BuildersDAO : BuildersDAOContainer.BuildersDAOBase
    {
        const string author = "REPLACE PLACEHOLDER HERE";

        //Step 1 - Implement Initialize Smart Contract Logic 
       public override Empty Initialize(Empty input) {}

        // Step 2 - Implement Join DAO Logic 
        public override Empty JoinDAO(Address input) {}

        // Step 3 - Implement Create Proposal Logic 
        public override Proposal CreateProposal(CreateProposalInput input) {}

        //Step 4 - Implement Vote on Proposal Logic 
        public override Proposal VoteOnProposal(VoteInput input) {}
        
        //Step 5 - Implement Get All Proposals Logic
        public override ProposalList GetAllProposals(Empty input) {}
     }    
  }
```
To begin, let’s first fill in the **author** string variable with a unique name.

:::warning
Aelf Sidechain does not recognise duplicate smart contracts hence we will be using the **author** variable as the unique identifier for our voting smart contract!
:::

>  as the aelf sidechain does not recognise duplicate smart contracts hence we will be using the **author** variable as the unique identifier for our voting smart contract!

Reviewer: ```suggestion
```

repeated above

**Implement Initialize Function**

Next, proceed to head over to the comment `//Step 1 - Implement Initialize Smart Contract Logic` where we will be implementing the Initialize functionality!

> In this function, we will first check if the smart contract has been already initialized and return if the smart contract has already been initialized!

Reviewer: ```suggestion
Within this function, our initial step involves verifying whether the smart contract has already been initialized. If it has, we'll promptly exit the function.
```
